Block syntax again

I've been working around variations of this for a while now,

````
c2hs: C header contains errors:

/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/usr/include/dispatch/object.h:143: (column 15) [ERROR]  >>> Syntax error !
  The symbol `^' does not fit here.
````

It happens when building the `OpenCL` package from hackage. It can be avoided by editing that file to simply remove the offending character while installing the package, then putting it back so as not to break anything. Is there anything we can do to either `c2hs` or `OpenCL` to avoid this?

--- a/c2hs.cabal
+++ b/c2hs.cabal
@@ -1,5 +1,5 @@
 Name:           c2hs
-Version:        0.25.2
+Version:        0.25.3
 License:        GPL-2
 License-File:   COPYING
 Copyright:      Copyright (c) 1999-2007 Manuel M T Chakravarty
@@ -100,6 +100,7 @@ flag base3
 
 Executable c2hs
     Build-Depends:  base >= 2 && < 5,
+                    bytestring,
                     language-c >= 0.4.7 && < 0.5,
                     filepath,
                     dlist

--- a/src/C2HS/C.hs
+++ b/src/C2HS/C.hs
@@ -65,6 +65,7 @@ import Language.C.Syntax
 import Language.C.Parser
 
 import Data.Attributes (Attr(..))
+import qualified Data.ByteString.Char8 as BS
 
 import C2HS.State  (CST,
                    fatal, errorsPresent, showErrors, raiseError,
@@ -95,6 +96,18 @@ parseHeader is pos =
       Left (ParseError (msgs,pos')) -> raiseError pos' msgs >>
                                        return (CTranslUnit [] undefNode)
       Right (ct,ns') -> setNameSupply ns' >> return ct
+
+-- | @bsReplace old new haystack@ replaces occurences of @old@ with
+-- @new@ in the @haystack@.
+bsReplace :: BS.ByteString -> BS.ByteString -> BS.ByteString -> BS.ByteString
+bsReplace old new = go id
+  where go acc hay
+          | BS.null hay = BS.concat (acc [])
+          | otherwise = case BS.breakSubstring old hay of
+                          (h,t) | BS.null t -> BS.concat (acc [h])
+                                | otherwise -> go (acc . (h:) . (new:))
+                                                  (BS.drop n t)
+        n = BS.length old
       
 -- | given a file name (with suffix), parse that file as a C header and do the
 -- static analysis (collect defined names)
@@ -108,7 +121,15 @@ loadAttrC fname  = do
                      -- read file
                      --
                      traceInfoRead fname
-                     contents <- liftIO (readInputStream fname)
+
+                     -- very hacky dodge of Apple's block syntax in
+                     -- type definitions. This simply replaces a block
+                     -- type with a function pointer type. The issue
+                     -- is that language-c does not support this
+                     -- syntax, but frameworks such as OpenCL now use
+                     -- it in their headers.
+                     contents <- fmap (bsReplace (BS.pack "(^") (BS.pack "(*"))
+                                      (liftIO $ readInputStream fname)
 
                      -- parse
                      --

